PROTOCOL.md

```diff
@@ -0,0 +1,519 @@
+- **Snapshot Isolation for Reads** - readers can read a consistent snapshot of a Delta table, even in the face of concurrent writes.
+- **Scalability to billions of partitions or file** - queries against a Delta table can be planned on a single machine or in parallel.
+- **Self describing** - all metadata for a Delta table is stored alongside the data, static tables can be copied or moved using standard tools.
+- **Support for incremental processing** - readers can tail the Delta log to determine what data has been added in a given period of time, allowing for efficient streaming.
+
+Delta's transactions are implemented using multi-version concurrency control (MVCC).
+As a table changes, Delta's MVCC algorithm  keeps multiple copies of the data around, rather than immediately replacing files that contain records that are being updated or removed.
+
+Readers of the table ensure that they only see one consistent _snapshot_ of a table at time by using the _transaction log_ to selectively choose which _data files_ to process.
+
+Writers modify the table in two phases:
+First, they optimistically write out new data files or updated copies of existing ones.
+Then, they _commit_, creating the latest _atomic version_ of the table by adding a new entry to the log.
+In this log entry they record which data files to logically add and remove, along with changes to other metadata about the table.
+
+Data files that are no longer present in the latest version of the table can be lazily deleted by the vacuum command, after a user-specified retention period (default 7 days).
+
+# Delta Table Specification
+A table has a single serial history of atomic versions, which are named using contiguous, monotonically-increasing integers.
+The state of a table at a given version is called a _snapshot_ and is defined by the following properties:
+ - **Version of the Delta log protocol** that is required to correctly read or write the table
+ - **Metadata** of the table (e.g., the schema, a unique identifier, partition columns, and other configuration properties)
+ - **Set of files** present in the table, along with metadata about those files
+ - **Set of tombstones** for files that were recently deleted
+ - **Set of applications-specific transactions** that have been successfully committed to the table
+
+## File Types
+A Delta table is stored within a directory and is composed of four different types of files.
+
+### Data Files
+Data files can be stored in the root directory of the table or in any non-hidden subdirectory (i.e., one whose name does not start with an `_`).
+By default, the reference implementation stores data files in directories that are named based on the partition values for data in that file (i.e. `part1=value1/part2=value2/...`).
+This directory format is only used to follow existing conventions and is not required by the protocol.
+Actual partition values for a file must be read from the transaction log.
```

Comment:
Should we add a TODO for how to parse partition values, such as convert it between string and the actual data type? One weird behavior in Spark and Delta is when the value of a partition column is an empty string, we will set its partition value to `null`. See [SPARK-24438](https://issues.apache.org/jira/browse/SPARK-24438)

Reply:
That is a great point.  I'll add this to the `Add FIle` action below.